### PR TITLE
Implement lazy-read and noshell raw mode

### PR DESCRIPTION
**Pull request by garazdawi**
This PR implements a couple of new fetures to how stdin and stdout is handled by Erlang.

* All reads from stdin are now done upon request, instead of greedingly reading all data. So it is only when an `io:get_line/2` or equivilant is issued by the Erlang program that a read is done. This solves #8113 and also the problem mentioned by Josè [here](https://github.com/erlang/otp/issues/8113#issuecomment-2088090720). This change also basically makes the `-noinput` flag obsolete, as no data will be read from stdin without it being requested.

* The `noshell` mode has been updated to have two "submodes" called `raw` or `cooked`, where `cooked` is the old default behaviour and `raw` can be used to bypass the line editing support of the native terminal. Using `raw` mode it is possible to read keystrokes as they happen without the user having to type Enter. The `raw` mode also does not echo the typed characters to stdout. An example of how to create a tic-tac-toe game using this mechanism is included in the documentation. This solves #8037.

* There is also a new guide added that shows how to create a custom shell with autocomplete using Erlang's builtin line editor (basically what Elixir, Luerl and other do).

### CODE REVIEW

**Comment by garazdawi** [comment]
You can read the new guides here:
*[Creating a custom shell](https://www.erlang.org/doc/apps/stdlib/custom_shell.html)

*[Creating a terminal application](https://www.erlang.org/doc/apps/stdlib/terminal_interface.html)

File: lib/stdlib/doc/guides/custom_shell.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+This guide will show how to create a custom shell. The most common
+use case for this is to support other languages running on the Erlang VM,
+but it can also be used to create specialized debugging shells a system.
+
+This guide will build on top of the built-in [Erlang line editor](`m:edlin`),
+which means that the keybindings described in [tty - A Command-Line Interface](`e:erts:tty.md`)
+can be used edit the input before it is passed to the custom shell. This
+somewhat limits what the custom shell can do, but it also means that we do not
+have to implement line editing ourselves. If you need more control over the
+shell, then use [Building a terminal interface](terminal_interface.md) as
```
Comment (zachallaun) [inline lib/stdlib/doc/guides/custom_shell.md:31]: Noticed this while reading through the new guides. 🙂 

```suggestion
shell, then use [Creating a terminal application](terminal_interface.md) as
```

**Comment by EliasDerHai** [comment]
> You can read the new guides here:
> 
>     * [Creating a custom shell](https://erlang.github.io/prs/8962/lib/stdlib-6.1.2/doc/html/custom_shell.html)
> 
>     * [Creating a terminal application](https://erlang.github.io/prs/8962/lib/stdlib-6.1.2/doc/html/terminal_interface.html)

broken links

**Comment by frazze-jobb** [comment]
They can now be found in the official documentation

* [Creating a custom shell](https://www.erlang.org/doc/apps/stdlib/custom_shell.html)

* [Creating a terminal application](https://www.erlang.org/doc/apps/stdlib/terminal_interface.html)

I updated the links above as well

**Comment by wmealing** [comment]
As a heads up. The "Creating a terminal application" links to https://www.erlang.org/doc/apps/stdlib/assets/tic-tac-toe.es, which has a problem.

The board characters served up the browser seem strangely encoded.
Screenshot 2025-11-04 at 1 53 00 pm

But when I use curl/wget it is encoded correctly.

Screenshot 2025-11-04 at 1 53 15 pm

I'll be looking where to report this correctly, but for now I'll report it here in case anyone trips over it.

**Comment by wmealing** [comment]
Probably a character encoding issue on the server.  

I think the problem is that its serving this file as: 

Content-Type:  application/ecmascript

While it should be:

Content-Type: text/html; charset=utf-8

**Pull request merged**